## Homework 1
Adam Lim, 4/21/19 <br></br>
Exercises from the course Learning From Data: https://work.caltech.edu/homeworks.html
<br></br>
Homework 1 link: https://work.caltech.edu/homework/hw1.pdf

### Question #1
What types of Machine Learning, if any, best describe the following three scenarios:<br></br>
- i. A coin classification system is created for a vending machine. The developers obtain exact coin specifications from the U.S. Mint and derive a statistical model of the size, weight, and denomination, which the vending machine then uses to classify coins.
- ii. Instead of calling the U.S. Mint to obtain coin information, an algorithm is presented with a large set of labeled coins. The algorithm uses this data to infer decision boundaries which the vending machine then uses to classify its coins.
- iii. A computer develops a strategy for playing Tic-Tac-Toe by playing repeatedly and adjusting its strategy by penalizing moves that eventually lead to losing.


<b> Answer </b> <br></br>
[d] (i) Not learning, (ii) Supervised Learning, (iii) Reinforcement Learning

### Question #2
Which of the following problems are best suited for Machine Learning?
- (i) Classifying numbers into primes and non-primes.
- (ii) Detecting potential fraud in credit card charges.
- (iii) Determining the time it would take a falling object to hit the ground.
- (iv) Determining the optimal cycle for traffic lights in a busy intersection.

<b> Answer </b> <br></br>
[a] (ii) and (iv)

### Question #3
We have 2 opaque bags, each containing 2 balls. One bag has 2 black balls and
the other has a black ball and a white ball. You pick a bag at random and
then pick one of the balls in that bag at random. When you look at the ball,
it is black. You now pick the second ball from that same bag. What is the
probability that this ball is also black?

<b> Answer </b> <br></br>
Bag 1 = [Black, Black] <br></br>
Bag 2 = [Black, White] <br></br>

Bayes Theorem: p(A|B) = P(B|A) * P(A) / P(B)

P(Ball 2 is Black | Ball 1 is Black) = p(Ball 1 is Black | Ball 2 is Black) * p(Ball 2 is black) / p(ball 1 is black)

In [1]:
(1* (0.5*0 + 0.5*1)) / (0.5*1 + 0.5*0.5)

0.6666666666666666

[d] 2/3

### Question #4
Consider a sample of 10 marbles drawn from a bin containing red and green marbles.
The probability that any marble we draw is red is µ = 0.55 (independently, with
replacement). We address the probability of getting no red marbles (ν = 0) in the
following cases:

We draw only one such sample. Compute the probability that ν = 0. The
closest answer is (‘closest answer’ means: |your answer−given option| is closest
to 0):

<b> Answer </b><br></br>

In [2]:
# Probability of sample having no red marbles
pow((1-0.55),10)

0.0003405062891601559

[b] 3.405 × (10 ^ −4)

### Question #5
We draw 1,000 independent samples. Compute the probability that (at least)
one of the samples has ν = 0. The closest answer is:

<b>Answer</b>
- P(1+ samples having no red) = 1 - P(All samples having red marbles)
- P(Sample having red marble) = 1 - P(Sample having no red marbles)


In [3]:
# P(Sample having red marble) = 1 - P(Sample having no red marbles)
p_red = 1 - pow((1-0.55),10)
print(p_red)

0.9996594937108398


In [4]:
# P(1+ samples having no red) = 1 - P(All samples having red marbles)
1 - pow(p_red, 1000)

0.28863119784980995

[c] 0.289

### Question #6
Consider a Boolean target function over a 3-dimensional input space X = {0, 1}
3
(instead of our ±1 binary convention, we use 0,1 here since it is standard for Boolean
functions). We are given a data set D of five examples represented in the table below,
where yn = f(xn) for n = 1, 2, 3, 4, 5. <br></br>

|   | Xn |   | Yn |
|---|----|---|----|
| 0 | 0  | 0 | 0  |
| 0 | 0  | 1 | 1  |
| 0 | 1  | 0 | 1  |
| 0 | 1  | 1 | 0  |
| 1 | 0  | 0 | 1  |


Note that in this simple Boolean case, we can enumerate the entire input space (since
there are only 2^3 = 8 distinct input vectors), and we can enumerate the set of all
possible target functions (there are only 2^2^3 = 256 distinct Boolean function on 3 Boolean inputs). <br></br><br></br>
Let us look at the problem of learning f. Since f is unknown except inside D, any
function that agrees with D could conceivably be f. Since there are only 3 points in
X outside D, there are only 2^3 = 8 such functions.
The remaining points in X which are not in D are: 101, 110, and 111. We want to
determine the hypothesis that agrees the most with the possible target functions. In
order to quantify this, count how many of the 8 possible target functions agree with
each hypothesis on all 3 points, how many agree on just 2 of the points, on just 1
point, and how many do not agree on any points. The final score for each hypothesis
is computed as follows:
<br></br><br></br>
Score = (# of target functions agreeing with hypothesis on all 3 points)×3 + (#
of target functions agreeing with hypothesis on exactly 2 points)×2 + (# of target
functions agreeing with hypothesis on exactly 1 point)×1 + (# of target functions
agreeing with hypothesis on 0 points)×0.

<b> Answer </b>
[e] They are all equivalent (equal scores for g in [a] through [d])

| Xn    |   |   | A  | B  | C  | D  |
|-------|---|---|----|----|----|----|
| 0     | 0 | 0 | 0  | 3  | 3  | 0  |
| 0     | 0 | 1 | 1  | 2  | 1  | 2  |
| 0     | 1 | 0 | 1  | 2  | 1  | 2  |
| 0     | 1 | 1 | 2  | 1  | 1  | 2  |
| 1     | 0 | 0 | 1  | 2  | 1  | 2  |
| 1     | 1 | 0 | 2  | 1  | 1  | 2  |
| 1     | 1 | 1 | 3  | 0  | 3  | 0  |
| 1     | 0 | 1 | 2  | 1  | 1  | 2  |
| Total |   |   | 12 | 12 | 12 | 12 |

### Questions #7 - #10
In this problem, you will create your own target function f and data set D to see
how the Perceptron Learning Algorithm works. Take d = 2 so you can visualize the
problem, and assume X = [−1, 1] × [−1, 1] with uniform probability of picking each
x ∈ X . <br></br><br></br>
In each run, choose a random line in the plane as your target function f (do this by
taking two random, uniformly distributed points in [−1, 1] × [−1, 1] and taking the
line passing through them), where one side of the line maps to +1 and the other maps
to −1. Choose the inputs xn of the data set as random points (uniformly in X ), and
evaluate the target function on each xn to get the corresponding output yn. <br></br><br></br>
Now, in each run, use the Perceptron Learning Algorithm to find g. Start the PLA
with the weight vector w being all zeros (consider sign(0) = 0, so all points are initially misclassified), and at each iteration have the algorithm choose a point randomly
from the set of misclassified points. We are interested in two quantities: the number
of iterations that PLA takes to converge to g, and the disagreement between f and
g which is P[f(x) != g(x)] (the probability that f and g will disagree on their classification of a random point). You can either calculate this probability exactly, or
approximate it by generating a sufficiently large, separate set of points to estimate it. <br></br><br></br>
In order to get a reliable estimate for these two quantities, you should repeat the
experiment for 1000 runs (each run as specified above) and take the average over
these runs.<br></br>

<b> Question #7: </b> Take N = 10. How many iterations does it take on average for the PLA to
converge for N = 10 training points? Pick the value closest to your results
(again, ‘closest’ means: |your answer − given option| is closest to 0). <br></br>

<b> Question #8: </b> Which of the following is closest to P[f(x) != g(x)] for N = 10?

<b>Question #9: </b> Now, try N = 100. How many iterations does it take on average for the PLA
to converge for N = 100 training points? Pick the value closest to your results.<br></br>

<b>Question #10: </b> Which of the following is closest to P[f(x) 6= g(x)] for N = 100?<br></br>

In [5]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
np.random.seed(99)

In [6]:
class RandomLine:
    def __init__(self, pt1, pt2):
        self.pt1 = pt1
        self.pt2 = pt2
        self.slope = (pt1[1] - pt2[1]) / (pt1[0] - pt2[0])
        # y - y1 = m(x - x1) -> y = mx - mx1 + y1 
        self.intercept = self.slope * (-1 * pt1[0]) + pt1[1]
    
    def evaluate_pt(self, random_pt):
        # Evaluate if point is above or below line
        line_pt = random_pt[0]*self.slope + self.intercept
        if random_pt[1] > line_pt:
            return 1
        else:
            return -1

In [7]:
### Initialize DF of N random points evaluated on specified random line
def PointsDF(N, line):
    pts_df = pd.DataFrame(np.random.uniform(-1,1,size=(N,2)))
    pts_df['x_vector'] = list(zip(np.ones(pts_df.shape[0]), pts_df[0], pts_df[1]))
    pts_df['x_vector'] = pts_df['x_vector'].apply(lambda x: np.asarray(x))
    pts_df['y_initial'] = pts_df['x_vector'].apply(lambda x: line.evaluate_pt(x[1:]))
    return pts_df

In [8]:
#### Evaluate using Perceptron Learning Algorithm
def converge(pts_df, N, f):
    weights = np.zeros(3) # Initialize vector of weights
    iters = 0

    pts_df['evaluate'] = pts_df['x_vector'].apply(lambda x: np.sign(np.dot(weights, x)))
    pts_df['match'] = np.where(pts_df.y_initial == pts_df.evaluate, 1, -1)
    matches = pts_df['match'].sum()
    
    while matches != N: # While 1+ misclassified points
        iters = iters + 1
        
        # Use misclassified to update weights vector
        temp = pts_df.query('match != 1').sample(1)
        weights = weights + temp['y_initial'].values[0]* temp['x_vector'].values[0]
        print("Iter {} : {}".format(iters, weights), file=f)

        pts_df['evaluate'] = pts_df['x_vector'].apply(lambda x: np.sign(np.dot(weights, x)))
        pts_df['match'] = np.where(pts_df.y_initial == pts_df.evaluate, 1, -1)
        matches = pts_df['match'].sum()

    print('Converged at Iteration {}'.format(iters), file=f)
    return iters, weights

In [9]:
N = 10
OOS_N = 1000

iters_list = []
probability_list = []

with open("HW1_output.txt", "a") as f:
    print('################## QUESTIONS 7 - 8 ##################', file=f)
    for i in range(0,1000):
        print('***************** TRIAL {} *****************'.format(i), file=f)

        # Initialize Random Points and Line
        line = RandomLine(np.random.uniform(-1,1, 2), np.random.uniform(-1,1, 2))
        pts_df = PointsDF(N, line)

        # Execute PLA
        iters, weights = converge(pts_df, N, f)

        # Evaluate Predictive ability on Out of Sample dataset
        OOS = PointsDF(OOS_N, line)
        OOS['Predict'] = OOS['x_vector'].apply(lambda x: np.sign(np.dot(weights, x)))
        OOS['Correct'] = np.where(OOS.Predict == OOS.y_initial, 1, 0)
        OOS_correct_rate = OOS['Correct'].sum() / OOS['Correct'].count()
        print(OOS_correct_rate, file=f)
        iters_list.append(iters)
        probability_list.append(OOS_correct_rate)

In [10]:
# Average number of iterations to converge
print(sum(iters_list)/len(iters_list))

# Average probability f(x) = g(x) in Out of Sample DF
print(sum(probability_list)/len(probability_list))

11.103
0.8894070000000001


<b> Answer (7) </b> <br></br>
[b] 15

<b> Answer (8) </b> <br></br>
[c] 0.1

In [11]:
N = 100
OOS_N = 1000

iters_list = []
probability_list = []

with open("HW1_output.txt", "a") as f:
    print('################## QUESTIONS 9 - 10 ##################', file=f)
    for i in range(0,1000):
        print('***************** TRIAL {} *****************'.format(i), file=f)

        # Initialize Random Points and Line
        line = RandomLine(np.random.uniform(-1,1, 2), np.random.uniform(-1,1, 2))
        pts_df = PointsDF(N, line)

        # Execute PLA
        iters, weights = converge(pts_df, N, f)

        # Evaluate Predictive ability on Out of Sample dataset
        OOS = PointsDF(OOS_N, line)
        OOS['Predict'] = OOS['x_vector'].apply(lambda x: np.sign(np.dot(weights, x)))
        OOS['Correct'] = np.where(OOS.Predict == OOS.y_initial, 1, 0)
        OOS_correct_rate = OOS['Correct'].sum() / OOS['Correct'].count()
        print(OOS_correct_rate, file=f)
        iters_list.append(iters)
        probability_list.append(OOS_correct_rate)

In [12]:
# Average number of iterations to converge
print(sum(iters_list)/len(iters_list))

# Average probability f(x) = g(x) in Out of Sample DF
print(sum(probability_list)/len(probability_list))

107.556
0.9866660000000032


<b> Answer (8) </b><br></br>
[b] 100

<b> Answer (10) </b><br></br>
[b] 0.01